### PR TITLE
ESQL: Add plan consistency verification after each optimizer

### PR DESCRIPTION
**Pull request by costin**
To catch potential mistakes in the output of a plan, introduce a sanity
 check rule for verifying the nodes dependency within a tree after each
 optimizer runs.

The goal of the rule is to assert that after being modified dependencies
 between nodes don't get misplaced or incorrectly replaced.

### CODE REVIEW

**Comment by costin** [comment]
The rule already exposed some problems in our plans that need addressing:
- [ ] The output of `FieldExtractExec` is inconsistent - it was modified as part of the spatial work by including `preferDocValues` however this is not included in the output - @craigtaverner 
- [ ] Grok / Dissect introduce new attributes with the same name without replacing those downstream in the plan creating problems
- [ ] Renaming the grouping field before a stats still relies on the original field - this might expose a bug in CombineProjections:
`from employees | rename gender AS foo | stats bar = count(*) by foo | drop foo | sort bar`

**Review by astefan — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java
Comment (luigidellaquila) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java:53]: At a first look I thought the usage of "generated" here was to account for intermediate values reused during the command calculation, eg. `EVAL a = 12, b = a + 2`

Looking closer I think it's not the case.

Debugging it I noticed that, for some commands, `references(p)` cannot distinguish between "used" and "emitted" values;
Eg. for MV_EXPAND `references(p)` returns both the expanding and the expanded attribute, that's why when you calculate the `missing` you have to remove the "generated".

This mechanism does not seem rock solid (ie. a command could pass this validation also declaring one of its needed inputs as a generated output), but I think it's a good approximation and a significant step forward in terms of plan validation.

**Review by luigidellaquila — APPROVED** [review]
I love it! LGTM

I left only one commend about the mechanics of the input validation (not sure it requires more clarifications, it took me some debugging to really understand the nuances of what's going on there).

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:101]: This runs the dependency check only after all optimization rules have run.

I think there's value in running the dependency check after each optimization rule, to be able pinpoint the offending optimization rule immediately - at least during tests, e.g. by using an assertion. What do you think?

This would probably require overriding `RuleExecutor.executeWithInfo`, so definitely out of scope for this PR.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java:38]: Hmm, this means we treat optimization errors like other verification failures - they become client (400) exceptions (`VerificationException`). I don't think this is correct.

Shouldn't we rather throw an `IllegalStateException` if the dependency check fails?

Also applies to the PhysicalVerifier.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java:43]: I'm confused about the naming here. I'd expect this class to contain all (or most) of our optimization rules, but they instead reside within the logical/physical optimizers' classes.

Maybe just `DependencyChecks`?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java:55]: nit: Error message specifically mentions optimization, so this class assumes it's only going to be used after optimizations. (Technically, dependencies should be fine also before any optimization has taken place, so we could check that, too.)
```suggestion
                failures.add(fail(p, "Plan [{}] invalid due to missing references {}", p.nodeString(), missing));
```

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java
Comment (alex-spies) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:3238]: I think we should also add a test for the case where a logical plan node in the middle of the plan has invalid references, not only the top node.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java
Comment (alex-spies) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java:53]: I agree that it's weird that for `MV_EXPAND x`, the expanded attribute for `x` will show up both in `generated` and in `refs` - maybe that's a bug in `MV_EXPAND`, and its `references` should not return the expanded attribute.

For `EVAL a = 12, b = a + 2` the constant folding takes care of the re-used `a`, but I found that for `from testidx | eval a = x +1, y = least(a, x)`, the re-used `a` shows up correctly in the refs _and_ the generated attributes.

**Comment by luigidellaquila** [comment]
> Grok / Dissect introduce new attributes with the same name without replacing those downstream in the plan creating problems

I'm wondering if it has something to do with [this](https://github.com/elastic/elasticsearch/issues/105043#issuecomment-1924298581).

**Review by alex-spies — COMMENTED** [review]
This is great!

Before merging, we need to make sure not to throw `VerificationException` (client error) in case of faulty optimizations, that should be some exception with 500 response code.

Otherwise, only minor remarks from me :)

**Review by fang-xing-esql — COMMENTED** [review]
LGTM in general. It is a good idea to validate the plan before sending it to downstream components as rail guard, however if we can avoid creating invalid plans at the first place that will be better, perhaps fixing the rules that create invalid plans are more complicated, and can be investigated later.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:101]: Not sure how useful that is considering there's a cost in setting and disabling the assertion - it will help localize the problem however as the code is stable I think it will be fairly easy to identify what happened without changing the rule executor.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java:38]: let's take this offline.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java
Comment (costin) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:1547]: It was a leftover and not used anywhere - a previous PR incorporated it yet the code hanged around - I've used this opportunity to remove it instead of creating another PR to further validate that the check gets applied.

**Comment by costin** [comment]
Since this PR is simple and was already approved but is blocked by 3 other issues, I've commented the rule and raised an issue to make sure it doesn't get forgotten.
The refactoring added are useful for #105425

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java
Comment (bpintea) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalVerifier.java]: Indeed, I think we should treat these planning inconsistencies different than throwing an `EsqlClientException`.

**Pull request merged**
